sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -616,21 +628,34 @@ public void close() throws IOException {
 
     abstract @Nullable String getPassword();
 
-    abstract Builder builder();
+    abstract @Nullable SerializableMessageMapper<EventT> getValueMapper();
+
+    abstract @Nullable SerializableFunction<EventT, String> getTopicNameMapper();
+
+    abstract @Nullable Coder<EventT> getCoder();
+
+    abstract Builder<EventT> builder();
 
     @AutoValue.Builder
-    abstract static class Builder {
-      abstract Builder setConnectionFactory(ConnectionFactory connectionFactory);
+    abstract static class Builder<EventT> {
+      abstract Builder<EventT> setConnectionFactory(ConnectionFactory connectionFactory);
+
+      abstract Builder<EventT> setQueue(String queue);
+
+      abstract Builder<EventT> setTopic(String topic);
+
+      abstract Builder<EventT> setUsername(String username);
 
-      abstract Builder setQueue(String queue);
+      abstract Builder<EventT> setPassword(String password);
 
-      abstract Builder setTopic(String topic);
+      abstract Builder<EventT> setValueMapper(SerializableMessageMapper<EventT> valueMapper);
```

Comment:
Thinking about the projects already using the legacy "setTopic" method, I think it would be a big breaking change. And introduces a big complexity as a lot of usage would not require dynamic topic use, but just setting the topic name